misc/peer-store/src/behaviour.rs

```diff
@@ -0,0 +1,122 @@
+use std::{collections::VecDeque, task::Poll};
+
+use libp2p_core::{Multiaddr, PeerId};
+use libp2p_swarm::{dummy, FromSwarm, NetworkBehaviour};
+
+use crate::store::Store;
+
+/// Events of this behaviour that will be emmitted to the swarm.
+pub enum Event {
+    RecordUpdated { peer: PeerId },
+}
+
+pub struct Behaviour<S> {
+    store: S,
+    /// Events that will be emitted.
+    pending_events: VecDeque<Event>,
+}
+
+impl<S> Behaviour<S>
+where
+    S: Store + 'static,
+{
+    pub fn new(store: S) -> Self {
+        Self {
+            store,
+            pending_events: VecDeque::new(),
+        }
+    }
+    /// List peers that are currently connected to this peer.
+    pub fn list_connected(&self) -> Box<[&PeerId]> {
+        self.store.list_connected()
+    }
+    /// Try to get all observed address of the given peer.  
+    /// Returns `None` when the peer is not in the store.
+    pub fn address_of_peer(&self, peer: &PeerId) -> Option<Box<[super::AddressRecord]>> {
+        self.store.addresses_of_peer(peer)
+    }
+    /// Manually update a record.  
+    /// This will always cause an `Event::RecordUpdated` to be emitted.
+    pub fn update_record(&mut self, peer: &PeerId, address: &Multiaddr) {
+        self.store.on_address_update(peer, address);
+        self.pending_events
+            .push_back(Event::RecordUpdated { peer: *peer });
+    }
+}
+
+impl<S> NetworkBehaviour for Behaviour<S>
+where
+    S: Store + 'static,
+{
+    type ConnectionHandler = dummy::ConnectionHandler;
+
+    type ToSwarm = Event;
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _connection_id: libp2p_swarm::ConnectionId,
+        peer: libp2p_core::PeerId,
+        _local_addr: &libp2p_core::Multiaddr,
+        remote_addr: &libp2p_core::Multiaddr,
+    ) -> Result<libp2p_swarm::THandler<Self>, libp2p_swarm::ConnectionDenied> {
+        self.store.on_peer_connect(&peer);
+        if self.store.on_address_update(&peer, remote_addr) {
+            self.pending_events.push_back(Event::RecordUpdated { peer });
+        }
+        Ok(dummy::ConnectionHandler)
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _connection_id: libp2p_swarm::ConnectionId,
+        peer: libp2p_core::PeerId,
+        _addr: &libp2p_core::Multiaddr,
+        _role_override: libp2p_core::Endpoint,
+        _port_use: libp2p_core::transport::PortUse,
+    ) -> Result<libp2p_swarm::THandler<Self>, libp2p_swarm::ConnectionDenied> {
+        self.store.on_peer_connect(&peer);

```

Comment:
It might make more sense to remove this for `FromSwarm` events since a connection could be denied later on (ie connection limits, banned peer, etc.), so that way the store isnt exactly cluttered.

Reply:
That makes sense. Will favor the swarm event.

Reply:
can we instead define Store::handle_* methods that are called here and in the other `NetworkBehaviour::handle_*` 
so that it allows us to further manage our peers?

Reply:
> can we instead define Store::handle_* methods that are called here and in the other `NetworkBehaviour::handle_*` so that it allows us to further manage our peers?

What are those specifically? The store no longer record connected peers.